src/datachain/lib/dc/datachain.py

```diff
@@ -712,10 +715,66 @@ def _resolve_checkpoint(
             and not checkpoints_reset
             and metastore.find_checkpoint(job.parent_job_id, _hash)
         ):
-            # checkpoint found → reuse dataset
+            # checkpoint found → find which dataset version to reuse
+
+            # Find dataset version that was created by any ancestor job
+            try:
+                dataset_version = metastore.get_dataset_version_for_job_ancestry(
+                    name,
+                    project.namespace.name,
+                    project.name,
+                    job.id,
+                )
+            except JobAncestryDepthExceededError:
+                raise JobAncestryDepthExceededError(
+                    "Job continuation chain is too deep. "
+                    "Please run the job from scratch without continuing from a "
+                    "parent job."
+                ) from None
+
+            if not dataset_version:
+                logger.debug(
+                    "Checkpoint found but no dataset version for '%s' "
+                    "in job ancestry (job_id=%s). Creating new version.",
+                    name,
+                    job.id,
+                )
+                # Dataset version not found (e.g deleted by user) - skip
+                # checkpoint and recreate
+                return _hash, None
+
+            logger.debug(
+                "Reusing dataset version '%s' v%s from job ancestry "
+                "(job_id=%s, dataset_version_id=%s)",
+                name,
+                dataset_version.version,
+                job.id,
+                dataset_version.id,
+            )
+
+            # Read the specific version from ancestry
             chain = read_dataset(
-                name, namespace=project.namespace.name, project=project.name, **kwargs
+                name,
+                namespace=project.namespace.name,
+                project=project.name,
+                version=dataset_version.version,
+                **kwargs,
             )
+
+            # Link current job to this dataset version (not creator) and update
+            # dataset_version.job_id atomically
+            with metastore.db.transaction() as conn:  # type: ignore[attr-defined]
```

Comment:
how would this transaction work in Studio? (the main concern) ... I'm not sure I understand this